Keep commits and translog up to the global checkpoint

dnhatn opened the pull request:
I have to include both index commit and translog policies in a single PR because they need one another for testing. This PR is a rework of https://github.com/elastic/elasticsearch/pull/27367.

We need to keep index commits and translog operations up to the current global checkpoint to allow us to throw away unsafe operations and increase the operation-based recovery chance. This is achieved by a new index deletion policy.

DaveCTurner commented on core/src/main/java/org/elasticsearch/common/lucene/index/ESIndexDeletionPolicy.java:
```diff
@@ -0,0 +1,48 @@
+     * @return a list of index commits that are not deleted by this policy.
+     */
+    List<IndexCommit> onCommit(List<? extends IndexCommit> commits) throws IOException;
+}
```
This interface clearly came from `IndexDeletionPolicy` but I think it's now different enough that it's worth breaking the link entirely and coming up with more descriptive names. The `onInit()` and `onCommit()` methods of `IndexDeletionPolicy` return `void` but here we're returning a list of the kept commits. Additionally, I think all the implementations just delegate `onInit()` to `onCommit()` (except `KeepUntilGlobalCheckpointDeletionPolicy` that special-cases an empty argument).

DaveCTurner commented on core/src/main/java/org/elasticsearch/common/lucene/index/ESIndexDeletionPolicy.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+/**
+ * An {@link IndexDeletionPolicy} that deletes unneeded index commits, and returns index commits are not deleted by this policy.
```
Not an `IndexDeletionPolicy` any more!

DaveCTurner commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -377,6 +359,28 @@ static SequenceNumbersService sequenceNumberService(
+            case OPEN_INDEX_AND_TRANSLOG:
+                final long globalCheckpoint = Translog.readGlobalCheckpoint(engineConfig.getTranslogConfig().getTranslogPath());
+                seqNoStats = store.loadSeqNoStats(globalCheckpoint);
+                break;
```
I'd just return `store.loadSeqNoStats()` here rather than `break`ing and `return`ing at the bottom. Not sure if that's against our style tho?

dnhatn replied:
Done

DaveCTurner commented on core/src/main/java/org/elasticsearch/index/engine/KeepUntilGlobalCheckpointDeletionPolicy.java:
```diff
@@ -0,0 +1,125 @@
+
+/**
+ * An {@link ESIndexDeletionPolicy} that deletes index commits that are not required for recovery.
+ * In particular, this policy will delete index commits whose max sequence number is smaller (or equal) than
```
`s/smaller (or equal) than/at most/` ?

dnhatn replied:
Done

DaveCTurner commented on core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java:
```diff
@@ -2516,11 +2547,8 @@ public void testSimpleCommit() throws IOException {
-        commit(translog, generation);
-    }
+        long lastGen = randomLongBetween(1, translog.currentFileGeneration());
+        commit(translog, randomLongBetween(1, lastGen), lastGen);    }
```
nit: `}` should be on its own line

dnhatn replied:
Done.

DaveCTurner commented on core/src/main/java/org/elasticsearch/index/engine/KeepUntilGlobalCheckpointDeletionPolicy.java:
```diff
@@ -0,0 +1,125 @@
+
+        final List<IndexCommit> keptCommits = new ArrayList<>();
+        final int keptPosition = indexOfKeptCommits(commits);
+        final List<Integer> duplicateIndexes = indexesOfDuplicateCommits(commits);
```
We only call this collection's `contains()` method so maybe a `HashSet<Integer>` would be better? Not sure if it gets big enough to make a difference.

dnhatn replied:
We expect to have zero or one element in a list, there should be no difference. However, using Set makes more sense here; I updated. Thank you.

DaveCTurner commented on core/src/test/java/org/elasticsearch/indices/recovery/RecoveryTests.java:
```diff
@@ -113,27 +113,10 @@ public void testRecoveryWithOutOfOrderDelete() throws Exception {
                 SourceToParse.source(orgReplica.shardId().getIndexName(), "type", "id2", new BytesArray("{}"), XContentType.JSON), u -> {});
 
-            final int translogOps;
+            final int translogOps = 4; // 3 ops + seqno gaps
```
This removes the tests for the cases where we've updated the index settings. Is that ok?

dnhatn replied:
I am working to have another test for this as the assumption in this test is no longer correct.

bleskes replied:
which other test is that?

ywelsch commented on core/src/main/java/org/elasticsearch/common/lucene/index/ESIndexDeletionPolicy.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * An {@link IndexDeletionPolicy} that deletes unneeded index commits, and returns index commits are not deleted by this policy.
+ */
+public interface ESIndexDeletionPolicy {
```
do we need a separate interface? If we want to know which commits can be kept,
we can just do `commits.stream().allMatch(c -> c.isDeleted() == false)` after invoking the `onInit` or `onCommit` methods?

dnhatn replied:
This interface is to avoid keeping translog of snapshotted index commits. `SnapshotDeletionPolicy` suppresses the `delete()` for snapshotted commits.

https://github.com/apache/lucene-solr/blob/e2521b2a8baabdaf43b92192588f51e042d21e97/lucene/core/src/java/org/apache/lucene/index/SnapshotDeletionPolicy.java#L221-L225

dnhatn replied:
I should have documented its purpose.

bleskes commented:
@dnhatn I have a question before diving deeper - I expected to find `CombinedDeletionPolicy` merged with `KeepUntilGlobalCheckpointDeletionPolicy` into a single class. The hunch was that that will removed the need to look at the index commits to know if there were deleted by another policy, allowing that information to be directly transferred to the translog deletion policy. Did that not work for some reason?

dnhatn commented:
I like the previous design, three policies are decoupled; I tried to keep them separated as before. I will embed `KeepUntilGlobalCheckpointDeletionPolicy` code to the CombinedDeletionPolicy.

dnhatn commented:
>  I expected to find CombinedDeletionPolicy merged with KeepUntilGlobalCheckpointDeletionPolicy into a single class

@bleskes, I have merged the `KeepUntilGlobalCheckpointDeletionPolicy` with `CombinedDeletionPolicy`. Could you please take a look? Thank you.

dnhatn commented:
@DaveCTurner I've addressed your comments but not all as I've removed the new interface.

bleskes commented:
> I like the current design, three policies are decoupled; I tried to keep them separated as before. I will embed KeepUntilGlobalCheckpointDeletionPolicy code to the CombinedDeletionPolicy.

I also find the composition of deletion policies elegant. Sadly elegance is an all or nothing thing. Adding extra wraps or copy interface means you lose it... 

> @bleskes, I have merged the KeepUntilGlobalCheckpointDeletionPolicy with CombinedDeletionPolicy. Could you please take a look? Thank you.

Thanks! I will look tomorrow morning.

DaveCTurner commented:
Yannick said:

> Do we need a separate interface? If we want to know which commits can be kept,
we can just do commits.stream().allMatch(c -> c.isDeleted() == false) after invoking the onInit or onCommit methods?

The interface has gone (👍) and the second half of this idea seems like a good one. The meaning of the return value of `deleteOldIndexCommits()` requires reading the comments/implementation as it's not clear from the name, whereas `commits.stream().allMatch(c -> c.isDeleted() == false)` is transparent.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -19,36 +19,45 @@
     @Override
     public void onInit(List<? extends IndexCommit> commits) throws IOException {
-        indexDeletionPolicy.onInit(commits);
+        final List<IndexCommit> keptCommits = deleteOldIndexCommits(commits);
```
can we rely on indexOfKeptCommits to give use the index of the first commit to be kept and use that to:
1) Delete all commits before
2) Set the translog deletion policy generation based on the commit the index points to.

dnhatn replied:
Yep, I agree. Let's remove the duplicate logic.

dnhatn commented:
@bleskes, There are some cases we may end up with multiple index commits having max_seqno.

1. If a (force) merge is executed, and the global checkpoint is still behind the max_seqno, we will have two commits with the same max_seqno. These two commits may require more disk space. When the global checkpoint catches up the max_seqno, we won't clean up the old commits unless a new commit is created. I added the duplicate logic to avoid this scenario.

2. A recovering commit from a local store or local shard is immediately replaced by a fresh commit with the same max_seqno. This is not a disk space issue as these two commits should share the same underlying files. The issue is the recovering does not have TRANSLOG_GEN in its commit which requires a special handling when processing translog generation. We can solve this without the duplicate logic.

https://github.com/elastic/elasticsearch/blob/2900e3f34528b21215ca736f65ae7fd5d75c5c80/core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L448-L468

bleskes commented:
> @bleskes, There are some cases we may end up with multiple index commits having max_seqno.

Thanks. Let's keep this a simple as possible. If those duplicate commit become a problem we can fix it later. Note though that when those duplicate commits are the "good ones", we should keep just one of them by choosing the commit appropriately. This is simple as it's just a question of select the commit index correctly.

dnhatn commented:
@bleskes, I've addressed your comments (ignore the CI failure). Could you please take another look? Thank you.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -21,34 +21,37 @@
     @Override
     public void onInit(List<? extends IndexCommit> commits) throws IOException {
-        indexDeletionPolicy.onInit(commits);
+        final int keptPosition = deleteOldIndexCommits(commits);
```
this is a question of taste, but I prefer of the deleteOldIndexCommits will return the oldest non-deleted commit (rather than it's index) and then we can pass two commits to the method connecting to the translog deletion policy: setLastCommittedTranslogGeneration(oldestCommit,lasterCommit)  . Feel free to keep as is if you like it better.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -66,24 +69,60 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
-        assert indexCommit.isDeleted() == false : "last commit is deleted";
-        long minGen = Long.parseLong(indexCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
-        translogDeletionPolicy.setMinTranslogGenerationForRecovery(minGen);
+    private void setLastCommittedTranslogGeneration(List<? extends IndexCommit> commits, final int keptPosition) throws IOException {
```
this method name is now obsolete. Maybe call it updateTranslogDeletionPolicy?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -66,24 +69,60 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
-    public TranslogDeletionPolicy getTranslogDeletionPolicy() {
-        return translogDeletionPolicy;
+    /**
+     * Find the index position of a safe index commit whose max sequence number is not greater than the global checkpoint.
```
Find the *highest* index position

bleskes commented on core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java:
```diff
@@ -76,6 +81,17 @@ public synchronized void setMinTranslogGenerationForRecovery(long newGen) {
+     * Sets the translog generation of the last index commit.
+     */
+    public synchronized void setTranslogGenerationOfLastCommit(long lastGen) {
+        if (lastGen < translogGenerationOfLastCommit) {
```
it would be nice to also check it's higher than the "required for recovery gen".

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java:
```diff
@@ -19,66 +19,136 @@
-            lastGen += randomIntBetween(10, 20000);
-            commitList.add(mockIndexCommitWithTranslogGen(lastGen));
+
+        int keptIndex = randomInt(commitList.size() - 2);
```
why can't we take the last one?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java:
```diff
@@ -19,66 +19,136 @@
+
+        globalCheckpoint.set(randomNonNegativeLong());
+        long safeTranslogGen = randomLongBetween(legacyTranslogGen, Long.MAX_VALUE);
+        IndexCommit safeCommit = mockIndexCommit(randomLongBetween(0, globalCheckpoint.get()), safeTranslogGen);
```
can we randomly first add non safe commit and then make it "safe"?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -946,17 +946,41 @@ public void testSearchResultRelease() throws Exception {
         assertThat(engine.getTranslog().currentFileGeneration(), equalTo(3L));
-        assertThat(engine.getTranslog().getDeletionPolicy().getMinTranslogGenerationForRecovery(), equalTo(3L));
+        assertThat(engine.getTranslog().getDeletionPolicy().getMinTranslogGenerationForRecovery(), equalTo(inSync.get() ? 3L : 1L));
+        assertThat(engine.getTranslog().getDeletionPolicy().getTranslogGenerationOfLastCommit(), equalTo(3L));
```
can we add a line that changes the inSync to true and checks that we cleaned up?

dnhatn replied:
Done.

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4089,4 +4130,63 @@ public void testSeqNoGenerator() throws IOException {
         }
     }
 
+    public void testKeepTranslogUpToGlobalCheckpoint() throws Exception {
```
I think you mean *after* the global checkpoint?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4089,4 +4130,63 @@ public void testSeqNoGenerator() throws IOException {
+                }
+                try (Translog.Snapshot snapshot = engine.getTranslog().newSnapshot()) {
+                    long requiredOps = engine.seqNoService().getMaxSeqNo() - Math.max(0, engine.seqNoService().getGlobalCheckpoint());
+                    assertThat("Should keep translog operations up to the global checkpoint",
```
can we assert we get all the ops above the global checkpoint in the snapshot?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4089,4 +4130,63 @@ public void testSeqNoGenerator() throws IOException {
+        final Path translogPath = createTempDir();
+        store = createStore();
+        try (InternalEngine engine
+                 = new InternalEngine(config(defaultSettings, store, translogPath, NoMergePolicy.INSTANCE, null), seqNoServiceSupplier)) {
```
can we randomize the translog retention policy?

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -1068,6 +1069,12 @@ public void testAcquireIndexCommit() throws IOException {
         commit.close();
+        // Make the global checkpoint in sync with the local checkpoint.
+        if (isPrimary) {
+            shard.getEngine().seqNoService().markAllocationIdAsInSync(shard.shardRouting.allocationId().getId(), numDocs + moreDocs - 1);
```
why not use org.elasticsearch.index.shard.IndexShard#updateLocalCheckpointForShard ?

dnhatn commented:
@bleskes, I've updated the deletion policy and addressed your comments (except the `RecoveryTests`). Can you take another look? Thank you!

dnhatn commented:
The PR is ready to review again. Please give it go @bleskes and @DaveCTurner. Thank you!

bleskes commented on core/src/test/java/org/elasticsearch/indices/recovery/RecoveryTests.java:
```diff
@@ -108,12 +114,21 @@ public void testRecoveryWithOutOfOrderDelete() throws Exception {
                 SourceToParse.source(orgReplica.shardId().getIndexName(), "type", "id", new BytesArray("{}"), XContentType.JSON),
                 u -> {});
 
+            if (isolateFirstBatch) {
```
we discussed an alternative simpler simulation.

dnhatn replied:
@bleskes, This test is more fun than we thought. Both `delete #1` and `index #0` are either kept or deleted at the same time.

1. To delete `delete #1` but keep `index #0`, we need isolate `delete #1` in its own translog `gen #1` (by flushing or rolling).

2. To keep `index #0` longer than `delete #1`, we need to store `index #0` with a higher seqno operation. However, this operation increases `max_seq`, then prevents the `delete 
 #0` from being deleted as we still have to keep the old commit (it's the only safe commit).

I used the size retention policy to keep `index #0` longer.

bleskes replied:
How about 
- delete #1 
- roll generation  (to create gen 2)
- index #0 
- index #3 
- flush (commit point has max seq no 3, and local checkpoint 1 -> points at gen 2, previous commit point is maintained)
- index #2 
- index #5 
- update the global checkpoint to #3
- tell lucene to delete unneeded commits so it will remove the old commit  (via IndexWriter#deleteUnusedFiles). Once we do a rollback on recovery, we can just flush again.

I think this will work?

dnhatn replied:
Yep, this works. I pushed https://github.com/elastic/elasticsearch/pull/27606/commits/6733b750d714955d2715eb1d67f8e31305e48b9e

dnhatn commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -21,43 +21,50 @@
+                // When an engine starts with OPEN_INDEX_CREATE_TRANSLOG, we can safely delete all existing index commits because:
+                // 1. The engine will create a fresh index commit with the same actual data immediately.
+                // 2. We should never refer these existing commits again as they belong to another engine.
+                commits.forEach(IndexCommit::delete);
```
Yes, we can safely delete all commits here. `IndexFileDeleter` keeps the last index commit's files[1] until we have a new commit even though the commit itself is removed the from the commit list. Lucene has a test[2] that deletes all commits in `onInit()`.

Anyway, I have updated the policy using `translogUUID` to delete invalid index commits. Let me know your thought @bleskes .

[1] https://github.com/apache/lucene-solr/blob/ea36f5040c3730d1e410eda76ad7ae3ecbbe68a1/lucene/core/src/java/org/apache/lucene/index/IndexFileDeleter.java#L240-L242
[2] https://github.com/apache/lucene-solr/blob/e2521b2a8baabdaf43b92192588f51e042d21e97/lucene/core/src/test/org/apache/lucene/index/TestDeletionPolicy.java#L508

bleskes replied:
Thanks this is very valuable for our future work.

dnhatn commented:
@bleskes, I have addressed your latest comments. Could you please take another look? Thank you.

dnhatn commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -380,6 +361,25 @@ static SequenceNumbersService sequenceNumberService(
+            case OPEN_INDEX_AND_TRANSLOG:
+                final long globalCheckpoint = Translog.readGlobalCheckpoint(engineConfig.getTranslogConfig().getTranslogPath());
+                return store.loadSeqNoStats(globalCheckpoint);
+
```
I pushed https://github.com/elastic/elasticsearch/pull/27606/commits/d8705b4c3a3e349484d6a0611900368e11a81e8e

dnhatn commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4149,10 +4149,19 @@ public long getGlobalCheckpoint() {
+            defaultSettings.getScopedSettings());
+        IndexMetaData.Builder builder = IndexMetaData.builder(indexSettings.getIndexMetaData());
+        builder.settings(Settings.builder().put(indexSettings.getSettings())
+            .put(IndexSettings.INDEX_TRANSLOG_RETENTION_AGE_SETTING.getKey(), randomFrom("-1", "10nanos", "500ms", "7s", "2m", "60m"))
```
I pushed https://github.com/elastic/elasticsearch/pull/27606/commits/d8705b4c3a3e349484d6a0611900368e11a81e8e

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -66,24 +74,55 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
-    public TranslogDeletionPolicy getTranslogDeletionPolicy() {
-        return translogDeletionPolicy;
+    /**
+     * Find the highest index position of a safe index commit whose max sequence number is not greater than the global checkpoint.
```
can we extend this to doc the first that we trim old commits that have a different translog uuid (and why)?

dnhatn replied:
I added a comment.

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4090,4 +4139,72 @@ public void testSeqNoGenerator() throws IOException {
         }
     }
 
+    public void testKeepTranslogAfterGlobalCheckpoint() throws Exception {
```
Do we need such a complicated test? can we just check that previous commit points are removed we open a new translog?

dnhatn replied:
I've simplified this test to verify that we keep only one safe commit.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -182,29 +179,13 @@ public InternalEngine(EngineConfig engineConfig) {
+                );
+                writer = createWriter(openMode == EngineConfig.OpenMode.CREATE_INDEX_AND_TRANSLOG);
                 updateMaxUnsafeAutoIdTimestampFromWriter(writer);
                 historyUUID = loadOrGenerateHistoryUUID(writer, engineConfig.getForceNewHistoryUUID());
```
can we assert that if we need to generate a new history uuid we always use `*_CREATE_TRANSLOG` as an open mode? that's we rely on the translog uuid only for trimming purposes (and avoid thinking about what it means to generate a new uuid)

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -4159,8 +4160,10 @@ public long getGlobalCheckpoint() {
-            .put(IndexSettings.INDEX_TRANSLOG_RETENTION_AGE_SETTING.getKey(), randomFrom("-1", "10nanos", "500ms", "7s", "2m", "60m"))
-            .put(IndexSettings.INDEX_TRANSLOG_RETENTION_SIZE_SETTING.getKey(), randomFrom("-1", "5b", "2kb", "1mb", "50gb"))
+            .put(IndexSettings.INDEX_TRANSLOG_RETENTION_AGE_SETTING.getKey(),
+                randomBoolean() ? "-1" : TimeValue.timeValueMillis(randomNonNegativeLong()).getStringRep())
```
Hmm..  not sure about this - I think it's very unlikely we'll end up with a small value that will cause rollovers. Wdyt?

bleskes commented on core/src/test/java/org/elasticsearch/indices/recovery/RecoveryTests.java:
```diff
@@ -120,26 +143,26 @@ public void testRecoveryWithOutOfOrderDelete() throws Exception {
-                    translogOps = 4; // 3 ops + seqno gaps
+                    translogOps = 6; // 5 ops + seqno gaps
                 }
+                orgReplica.updateGlobalCheckpointOnReplica(1L, "test");
```
add a comment as to why?

bleskes commented on core/src/test/java/org/elasticsearch/indices/recovery/RecoveryTests.java:
```diff
@@ -120,26 +143,26 @@ public void testRecoveryWithOutOfOrderDelete() throws Exception {
+                    translogOps = 6; // 5 ops + seqno gaps
                 }
+                orgReplica.updateGlobalCheckpointOnReplica(1L, "test");
                 flushShard(orgReplica);
```
I'm not sure about this - it also creates a new commit point and that is the one we'll use for recovery (because we don't do rollback yet) and it will point at the last translog gen, I think. Are you sure it's ok?

dnhatn commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -21,43 +21,51 @@
                 assert commits.isEmpty() == false : "index is opened, but we have no commits";
+                // When an engine starts with OPEN_INDEX_CREATE_TRANSLOG, we can delete all commits except the last commit,
+                // without updating the translog policy because these commits do not belong this engine.
+                for (int i = 0; i < commits.size() - 1; i++) {
```
Yep, I've removed this 👍

dnhatn commented:
@bleskes, I've added an assertion for `forceNewHistoryUUID` and updated the recovery test as we discussed. Could you please take another look? Thank you.

dnhatn commented:
@elasticmachine please test this.

dnhatn commented:
@DaveCTurner Could you please have a look? Thank you.

dnhatn commented:
Thanks @bleskes and @DaveCTurner.
